src/ServiceControl.LoadTests.AuditGenerator/Program.cs

```diff
@@ -1,115 +1,397 @@
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
     using System.Threading;
     using System.Threading.Tasks;
-    using Messages;
-    using Metrics;
     using NServiceBus;
+    using NServiceBus.Extensibility;
+    using NServiceBus.Features;
+    using NServiceBus.Routing;
     using NServiceBus.Support;
+    using NServiceBus.Transport;
+    using ServiceControl.Infrastructure.Metrics;
     using Transports;
 
     class Program
     {
-        const string ConfigRoot = "AuditGenerator";
+        public static readonly string HostId = Guid.NewGuid().ToString();
 
-        static void Main()
+        public static IDispatchMessages Dispatcher { get; set; }
+
+        static async Task Main(string[] commandLineArgs)
         {
-            Start().GetAwaiter().GetResult();
+            bodySize = int.Parse(commandLineArgs[0]);
+            var transportCustomizationName = commandLineArgs.Length > 1
+                ? commandLineArgs[1]
+                : QueueLengthProviderMsmqTransportCustomizationName;
+
+            var connectionString = commandLineArgs.Length > 2 ? commandLineArgs[2] : null;
+
+            var transportCustomization = (TransportCustomization)Activator.CreateInstance(Type.GetType(transportCustomizationName, true));
+            var queueLengthProvider = transportCustomization.CreateQueueLengthProvider();
+            queueLengthProvider.Initialize(connectionString, CacheQueueLength);
+
+            var configuration = new EndpointConfiguration("AuditGenerator");
+            configuration.EnableFeature<MyDispatchHackFeature>();
+            configuration.SendOnly();
+            configuration.UseTransport<MsmqTransport>();
+            configuration.UsePersistence<InMemoryPersistence>();
+            configuration.EnableInstallers();
+
+            GenerateLayouts();
+
+            var metrics = new Metrics();
+            reporter = new MetricsReporter(metrics, Console.WriteLine, TimeSpan.FromSeconds(2));
+            counter = metrics.GetCounter("Messages sent");
+
+            var endpoint = await Endpoint.Start(configuration);
+
+            var commands = new (string, Func<CancellationToken, string[], Task>)[]
+            {
+                ("f|Fill the sender queue. Syntax: f <number of messages> <number of tasks> <destination>",
+                    (ct, args) => Fill(args, Dispatcher)),
+                ("s|Start sending messages to the queue. Syntax: s <number of tasks> <destination>",
+                    (ct, args) => FullSpeedSend(args, Dispatcher, ct)),
+                ("t|Throttled sending that keeps the receiver queue size at n. Syntax: t <number of msgs> <destination>",
+                    (ct, args) => ConstantQueueLengthSend(args, Dispatcher, queueLengthProvider, ct)),
+                ("c|Constant-throughput sending. Syntax: c <number of msgs per second> <destination>",
+                    (ct, args) => ConstantThroughputSend(args, Dispatcher, ct))
```

Comment:
One thing to keep in mind is that we had various degrees of success with these options in the previous TF. They felt pretty unstable in their behavior. We fixed one (I don't recall which one) and used that only. @danielmarbach, do you remember which option we used to load test RavenDB 5?

Reply:
Throttled I think. The constant one overloads things

Reply:
I think we can remove them.

- **f** might be useful to do a quick smoke test but we can just run **t** for a while to do the same job.
- **s** will pretty much always overwhelm the system
- **c** if we set it too high then the SC instance will get overwhelmed and then it's functionally not different to **s**. If we set it too low then the SC instance will spend a bunch of time idle. This one _might_ be useful if ServiceControl ever does any sort of adaptive processing (i.e. turning off features to dedicate more resources to ingestion if the queue length gets high enough). Even then, we'd probably be more interested in simulating waves of peak processing.

Reply:
Decided to keep c and t